zerver/lib/user_groups.py

```diff
@@ -192,25 +192,103 @@ def access_user_group_for_update(
+
+
+def remove_user_group_from_group_based_setting(
+    object_with_settings: Stream | NamedUserGroup | Realm,
+    setting_name: str,
+    user_group: NamedUserGroup,
+) -> tuple[int | UserGroupMembersData, int | UserGroupMembersData]:
+    old_user_group = getattr(object_with_settings, setting_name)
+    old_setting_api_value = get_group_setting_value_for_api(old_user_group)
+    if isinstance(old_setting_api_value, int):
+        assert old_setting_api_value == user_group.id
+        return (
+            get_replacement_value_for_group_based_setting(object_with_settings, setting_name).id,
+            old_setting_api_value,
+        )
+
+    assert isinstance(old_setting_api_value, UserGroupMembersData)
+    assert user_group.id in old_setting_api_value.direct_subgroups
+
+    # Construct the new setting value by dropping our user_group from the list of subgroups.
+    new_subgroups_value = list(
+        filter(
+            lambda subgroup_id: subgroup_id != user_group.id, old_setting_api_value.direct_subgroups
+        )
+    )
+    new_setting_value = UserGroupMembersData(
+        direct_members=old_setting_api_value.direct_members, direct_subgroups=new_subgroups_value
+    )
+    # parse_group_setting_value will ensure to return a sane final setting
+    # value, avoiding the formation of a pathological anonymous group, and
+    # replacing it with the replacement group if it was about to become empty.
+    parsed_new_setting_value = parse_group_setting_value(
+        new_setting_value,
+        nobody_group=get_replacement_value_for_group_based_setting(
+            object_with_settings, setting_name
+        ),
+    )
+
+    return parsed_new_setting_value, old_setting_api_value
+
+
+@transaction.atomic(savepoint=False)
+def check_user_group_can_be_deactivated(
+    user_group: NamedUserGroup, *, force: bool = False
+) -> list[dict[str, Any]]:
     """
-    Main security check / access function for whether the acting
-    user has permission to deactivate a given user group.
+    This function is called in advance of deactivating a group.
+    Ordinarily (with force=False), checks for the conditions that prevent
+    group deactivation - returns a list of objections if such conditions
+    are present.
+    With force=False, it forcefully resolved such issues, in order to prepare
+    the group for correct deactivation.
+
+    The conditions we check and resolve are:
+
+    1. Is user_group a subgroup of any supergroups?
+       With force=True, remove it from such supergroups.
+    2. Are there any permissions that rely on the user group?
+       This has two possible cases:
+       a) A permission might be set directly to the group.
+          With force=True, we set such a permission to its corresponding
+          replacement_group_name.
+       b) A permission might be set to an anonymous group, which has user_group
+          as a subgroup.
+          With force=True, we remove user_group from the anonymous group.
```

Comment:
Should a function that has `check` at it's prefix modify data if the force flag is true? 

I haven't studied the code in this function deeply, so checking if we can have a different function to prep a group for removal if needed instead of that logic being inside the `check` function? 

The `if force` and `if not force` also makes the function bit hard to follow.

Reply:
I imagine it might be possible if we get a meaningful list of objections? Not too sure.

Reply:
> Should a function that has check at it's prefix modify data if the force flag is true?

@shubham-padia Yeah, I don't love it. It's not unprecedented - e.g. the `check_send_..._message` functions actually send a message, without even any flag.

Ideally I'd prefer a separate function, e.g. `force_prepare_group_for_deactivation` for doing this, instead of the `if not force: <build objections> elif force: <force-fix issues>` branching. But such a function would have to completely duplicate all the logic of `check_user_group_can_be_deactivated` of finding supergroups and permissions. I figured that's going to be very fragile, since this logic is relatively complex

Reply:
Got it, makes sense in that case. We can refactor later if we find a better way.